src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,82 +700,199 @@ namespace nbl::ext::imgui
-				assert(memOffset.isValid());
-			}
+				retV.scale = core::vector2df_SIMD{ 2.0f / drawData->DisplaySize.x , 2.0f / drawData->DisplaySize.y };
+				retV.translate = core::vector2df_SIMD { -1.0f, -1.0f } - core::vector2df_SIMD{ drawData->DisplayPos.x, drawData->DisplayPos.y } * trs.scale;
 
-			IGPUBuffer::SCreationParams indexCreationParams = {};
-			indexCreationParams.usage = nbl::core::bitflag(nbl::asset::IBuffer::EUF_VERTEX_BUFFER_BIT) | nbl::asset::IBuffer::EUF_INDEX_BUFFER_BIT
-				| nbl::asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
-			indexCreationParams.size = indexSize;
+				return std::move(retV);
+			}();
 
-			auto & indexBuffer = m_indexBuffers[frameIndex];
+			static constexpr auto MDI_ALLOCATION_COUNT = MDI::EBC_COUNT;
+			static constexpr auto MDI_ALIGNMENTS = std::to_array<typename MDI::COMPOSE_T::size_type, MDI_ALLOCATION_COUNT>({ alignof(VkDrawIndexedIndirectCommand), alignof(PerObjectData), alignof(ImDrawIdx), alignof(ImDrawVert) });
 
-			if (static_cast<bool>(indexBuffer) == false || indexBuffer->getSize() < indexSize)
+			struct MULTI_ALLOC_PARAMS
 			{
-				indexBuffer = m_device->createBuffer(std::move(indexCreationParams));
+				std::array<typename MDI::COMPOSE_T::size_type, MDI_ALLOCATION_COUNT> byteSizes = {};
+				std::array<typename MDI::COMPOSE_T::value_type, MDI_ALLOCATION_COUNT> offsets = {};
+			};
+
+			MULTI_ALLOC_PARAMS multiAllocParams;
+			std::fill(multiAllocParams.offsets.data(), multiAllocParams.offsets.data() + MDI_ALLOCATION_COUNT, MDI::COMPOSE_T::invalid_value);
+			
+			// calculate upper bound byte size for each mdi's content
+			for (uint32_t i = 0; i < drawData->CmdListsCount; i++)
+			{
+				const ImDrawList* commandList = drawData->CmdLists[i];
 
-				video::IDeviceMemoryBacked::SDeviceMemoryRequirements memReq = indexBuffer->getMemoryReqs();
-				memReq.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
-				auto memOffset = m_device->allocate(memReq, indexBuffer.get());
-				assert(memOffset.isValid());
+				multiAllocParams.byteSizes[MDI::EBC_DRAW_INDIRECT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(VkDrawIndexedIndirectCommand);
+				multiAllocParams.byteSizes[MDI::EBC_ELEMENT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(PerObjectData);
+				multiAllocParams.byteSizes[MDI::EBC_INDEX_BUFFERS] += commandList->IdxBuffer.Size * sizeof(ImDrawIdx);
+				multiAllocParams.byteSizes[MDI::EBC_VERTEX_BUFFERS] += commandList->VtxBuffer.Size * sizeof(ImDrawVert);
 			}
 
+			// calculate upper bound byte size limit for mdi buffer
+			const auto mdiBufferByteSize = std::reduce(std::begin(multiAllocParams.byteSizes), std::end(multiAllocParams.byteSizes));
+
+			auto mdiBuffer = smart_refctd_ptr<IGPUBuffer>(m_mdi.streamingTDBufferST->getBuffer());
 			{
-				auto vBinding = vertexBuffer->getBoundMemory();
-				auto iBinding = indexBuffer->getBoundMemory();
+				std::chrono::steady_clock::time_point timeout(std::chrono::seconds(0x45));
```

Comment:
timeout waaay too long, should be like max 1 second

Reply:
in fact you don't have a timeout at all!

This is setting the timeout point to 1970!!!!!